File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    void bfs(int i, int j, vector<vector<int>>& visited,
             vector<vector<char>>& grid) {
        visited[i][j] = 1;

        queue<pair<int, int>> q;
        q.push({i, j});

        int n = grid.size();
        int m = grid[0].size();

        int drow[] = {-1, 1, 0, 0};
        int dcol[] = {0, 0, -1, 1};

        while (!q.empty()) {
            int row = q.front().first;
            int col = q.front().second;
            q.pop();

            for (int i = 0; i < 4; i++) {
                int nrow = row + drow[i];
                int ncol = col + dcol[i];

                if (nrow >= 0 && ncol >= 0 && nrow < n && ncol < m &&
                    !visited[nrow][ncol] && grid[nrow][ncol] == '1') {

                    visited[nrow][ncol] = 1;
                    q.push({nrow, ncol});
                }
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int ans = 0;
        int n = grid.size();
        int m = grid[0].size();

        vector<vector<int>> visited(n, vector<int>(m, 0));

        for (int row = 0; row < n; row++) {
            for (int col = 0; col < m; col++) {
                if (!visited[row][col] && grid[row][col] == '1') {
                    ans += 1;
                    bfs(row, col, visited, grid);
                }
            }
        }
        return ans;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public void bfs(int mm,int nn,char[][] grid,List<List<Boolean>> visited){
        visited.get(mm).set(nn,true);

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{mm,nn});

        int m = grid.length;
        int n = grid[0].length;

        int[] drow = {1,-1,0,0};
        int[] dcol = {0,0,1,-1};

        while(!q.isEmpty()){
            int[] cell = q.poll();
            int row = cell[0];
            int col = cell[1];

            for(int i=0;i<4;i++){
                int nrow = row + drow[i];
                int ncol = col + dcol[i];


                if(nrow >= 0 && ncol >= 0 && nrow < m && ncol < n && grid[nrow][ncol] == '1' && !visited.get(nrow).get(ncol)){
                    visited.get(nrow).set(ncol,true);
                    q.add(new int[]{nrow,ncol});
                }
            }
        }
    }
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;

        List<List<Boolean>> visited = new ArrayList<>();

        for(int i = 0;i<m;i++){
            List<Boolean> row = new ArrayList<>();
            for(int j = 0;j<n;j++){
                row.add(false);
            }

            visited.add(row);
        }

        int ans = 0;

        for(int i = 0;i<m;i++){
            for(int j = 0;j<n;j++){
                if(!visited.get(i).get(j) && grid[i][j] == '1'){
                    ans++;
                    bfs(i,j,grid,visited);
                }
            }
        }

        return ans;
    }
}